libtenzir/builtins/operators/grok.cpp

```diff
@@ -0,0 +1,517 @@
+        result_pattern.append(replacement);
+      }
+      // We'll also have all the same named captures as the subpattern,
+      // except if they have a name that we already have saved:
+      // we don't want subpattern captures to overwrite anything we have in the
+      // main pattern.
+      //
+      // We expect `named_captures` to be quite small, so O(n) lookup is okay
+      std::ranges::copy_if(subpattern.named_captures,
+                           std::back_inserter(named_captures),
+                           [&](const auto& capture) {
+                             return get_duplicate_capture_name(capture.first)
+                                    == named_captures.end();
+                           });
+      previous_match_end = replacement_field.end();
+    }
+    result_pattern.append(previous_match_end, str_re2.end());
+    resolved_pattern.emplace(result_pattern, boost::regex_constants::no_except);
+    if (resolved_pattern->empty())
+      diagnostic::error("invalid regular expression")
+        .hint("regex: `{}`", result_pattern)
+        .throw_();
+  }
+}
+
+void pattern_store::parse_line(std::string_view line) {
+  if (line.empty())
+    return;
+  if (line.starts_with('#'))
+    return;
+  auto parts = detail::split(line, " ", 1);
+  TENZIR_ASSERT(not patterns.contains(std::string{parts[0]}));
+  patterns.emplace(std::string{parts[0]}, std::string{parts[1]});
+}
+
+auto& get_builtin_pattern_store() {
+  // Using CoW to avoid copying the builtin patterns unless we need to
+  static auto store = caf::make_copy_on_write<pattern_store>(
+    std::span{builtin_patterns_strings});
+  return store;
+}
+
+class grok_parser final : public plugin_parser {
+public:
+  grok_parser() = default;
+
+  explicit grok_parser(parser_interface& p)
+    : patterns_(get_builtin_pattern_store()) {
+    auto parser
+      = argument_parser{"grok", "https://docs.tenzir.com/next/operators/"
+                                "transformations/grok"};
+    parser.add(input_pattern_.raw_pattern, "<input_pattern>");
+    std::optional<std::string> pattern_definitions{};
+    parser.add("--pattern-definitions", pattern_definitions, "<patterns>");
+    parser.add("--indexed-captures", indexed_captures_);
+    parser.add("--include-unnamed", include_unnamed_);
+    parser.parse(p);
+    if (pattern_definitions)
+      patterns_.unshared().add(*pattern_definitions);
+    TENZIR_ASSERT_EXPENSIVE(std::ranges::all_of(
+      patterns_->patterns | std::views::values, [](const auto& p) -> bool {
+        return p.resolved_pattern.has_value();
+      }));
+    input_pattern_.resolve(*patterns_, false);
+  }
+
+  auto name() const -> std::string override {
+    return "grok";
+  }
+
+  auto
+  instantiate(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> std::optional<generator<table_slice>> override {
+    (void)input;
+    diagnostic::error("`{}` cannot be used here", name())
+      .emit(ctrl.diagnostics());
+    return {};
```

Comment:
I expected this to just work line-by-line, returning `tenzir.grok` events. 

If `parse_strings` takes a `generator<string_view>` and you do the `values(string_type{}, *arr)` in the calling context for it, then you basically get this for free.

Reply:
This piece is actually copied from the `kv` PR: https://github.com/tenzir/tenzir/pull/3646

If we want to make the design decision that `grok` is usable outside of `parse` as a standalone operator, I can make that change.

Reply:
`kv` will have to be adapted in the same way. I think this change makes sense, so +1 from me :+1:

Reply:
I'd be fine with having it in a separate PR, just generally wanted to note it down. I think we should do that before including the two parsers in a release. Otherwise we'd have to document that the parsers are not usable with the `read` operator.